src/font.rs

```diff
@@ -264,7 +264,13 @@ impl Font {
         }
 
         if let Some(features) = self.features.as_ref() {
-            fs::write(path.join(FEATURES_FILE), features)?;
+            // Normalize feature files with line feed line endings
+            // This is consistent with the line endings serialized in glif and plist files
+            if features.contains('\r') {
```

Comment:
just a thought, and maybe the compiler is smart enough to optimize this, but I would prefer to scan for the `\r` *byte* over the `\r` *char*; scanning chars means decoding utf-8, whereas scanning bytes doesn't.

Reply:
My only other thought (and this is totally unnecessary) is that if we really wanted to be perf nerds, the better approach would probably be to do this replacement when reading, instead of when writing.

Basically:

- instead of something like `fs::read_to_string`, manually open the fea as a `File`
- allocate two buffers
- read chunks from the file into the first buffer. on each read, scan the chunk for `\r`
- if `\r` is encountered, then manually recopy the lines into the second buffer, skipping an `\r`
- if `\r` is never encountered, just return the first buffer untouched as the result, else return the second buffer.


this is dumb and definitely not worth it. 🤷

Reply:
> scanning chars means decoding utf-8

I was wondering about the read approach when I worked on this.  It looks like the feature read is to String so the UTF-8 check happens read side.  


```rust
fn load_features(features_path: &Path) -> Result<String, Error> {
    let features = fs::read_to_string(features_path)?;
    Ok(features)
}
```

Does the feature file need to be a UTF-8 vetted String?  On the one hand, I suppose that it provides a built-in linting check at read time.  On the other, it may not be necessary unless there are String'y things that will need to be done with it in the library.  Right now it appears that feature data are exposed as is to users in a String.

Reply:
> I would prefer to scan for the \r byte over the \r char

Would this be something along the lines of searching for `b"\r"` in `features.as_bytes()`?

Reply:
Exactly.

Reply:
> > scanning chars means decoding utf-8

what I mean is simply that the chars() iter has to determine character boundaries in the underlying bytes, which has some overhead. This is on top of the overhead of validating the utf-8 when it's first read. 
> 
> I was wondering about the read approach when I worked on this.  It looks like the feature read is to String so the UTF-8 check happens read side.

Yes, I believe we read directly to string currently, which handles validation. We could alternatively read to bytes and then validate ourselves, to equal effect. 
> 
> 
> 
> 
> 
> ```rust
> 
> fn load_features(features_path: &Path) -> Result<String, Error> {
> 
>     let features = fs::read_to_string(features_path)?;
> 
>     Ok(features)
> 
> }
> 
> ```
> 
> 
> 
> Does the feature file data need to be a UTF-8 vetted String?  On the one hand, I suppose that it provides a built-in encoding linting check at read time.  On the other, it may not be necessary unless there are String'y things that will need to be done with it in the library.  Right now it appears that feature data are exposed as is to users in a String without other feature support.

This is a good question, which is not answered by the spec. The spec does mention that strings are in utf-8, and I would prefer to assume that the whole file is, as well. It might be worth opening an issue in the Adobe repo...

Reply:
converted to a byte check in 2e84bce7ccd126a9fc8b9a3f5338e067eed25139

Reply:
> This is a good question, which is not answered by the spec. The spec does mention that strings are in utf-8, and I would prefer to assume that the whole file is, as well. It might be worth opening an issue in the Adobe repo...

Response from Josh Hadley (Adobe):

> Yes, FEA files should indeed use UTF-8 encoding throughout. This has been an unwritten (well…partially written) assumption for some time now but as of the latest AFDKO (3.7.1), that assumption is enforced by the new Antlr4-based parser in makeotfexe. Older versions of makeotfexe might have been lax on that point, but that will no longer be the case moving forward. Also worth noting that makeotf (the Python interface to makeotfexe) has effectively enforced this for a while already.
>
> But as you note, the OpenType Feature File Specification does not actually explicitly state that the whole file should be UTF-8. We will clarify that in an update soon.